axelor-sale/src/main/java/com/axelor/apps/sale/service/saleorder/SaleOrderLineTreeServiceImpl.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Axelor Business Solutions
+ *
+ * Copyright (C) 2005-2023 Axelor (<http://axelor.com>).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package com.axelor.apps.sale.service.saleorder;
+
+import com.axelor.apps.base.AxelorAlertException;
+import com.axelor.apps.base.db.Product;
+import com.axelor.apps.base.db.repo.ProductRepository;
+import com.axelor.apps.sale.db.*;
+import com.axelor.apps.sale.db.repo.*;
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.List;
+import org.apache.commons.collections.CollectionUtils;
+
+public class SaleOrderLineTreeServiceImpl implements SaleOrderLineTreeService {
+  protected SaleOrderLineTreeRepository saleOrderLineTreeRepository;
+  protected ProductRepository productRepository;
+
+  protected SaleOrderLineRepository saleOrderLineRepository;
+  BigDecimal price = new BigDecimal(0);
+  BigDecimal subTotalCostPrice = new BigDecimal(0);
+
+  @Inject
+  public SaleOrderLineTreeServiceImpl(
+      SaleOrderLineTreeRepository saleOrderLineTreeRepository,
+      ProductRepository productRepository,
+      SaleOrderLineRepository saleorderLineRepository) {
+    this.saleOrderLineTreeRepository = saleOrderLineTreeRepository;
+    this.productRepository = productRepository;
+    this.saleOrderLineRepository = saleorderLineRepository;
+  }
+
+  @Override
+  public SaleOrderLineTree createLineTree(Product product) {
+
+    List<Product> productComponents =
+        productRepository.all().filter("self.parentProduct.id = ?1", product.getId()).fetch();
+
+    SaleOrderLineTree tree = null;
+    if (!productComponents.isEmpty()) {
+      tree = new SaleOrderLineTree();
+      tree.setCode(product.getCode());
+      tree.setTitle(product.getName());
+      tree.setDescription(product.getDescription());
+      tree.setParent(createLineTree(product.getParentProduct()));
+      tree.setProduct(product);
+
+      tree.setUnit(product.getUnit());
+      tree.setUnitCost(product.getCostPrice());
+      tree.setUnitPrice(product.getSalePrice());
+      tree.setTotalCost(product.getCostPrice().multiply(tree.getQuantity()));
+      tree.setTotalPrice(product.getSalePrice().multiply(tree.getQuantity()));
```

Comment:
declare following variables to avoid multiple calls to the same getters:
* costPrice
* salePrice
* quantity

please do the same in the rest of your changes